### ISSUE TITLE
[NNX] TernaryDense — ternary-weight dense layer for BitNet b1.58 style models

### ISSUE THREAD

**Issue by simeon-kepp** [google/flax#5468]
## Summary

Proposing a `TernaryDense` layer for NNX as a drop-in replacement for `nn.Linear` in ternary-weight models (BitNet b1.58 and successors).

## Motivation

BitNet b1.58 ([arxiv 2402.17764](https://arxiv.org/abs/2402.17764)) showed that training with weights quantized to `{-1, 0, +1}` reaches competitive performance vs full-precision at scale, with significant inference savings at high sparsity (55–65% zero weights). Several follow-on works use the same approach. Community interest is growing as inference cost becomes a primary concern.

The main things a JAX/Flax user needs to run these models that aren't currently in the library:

1. **A trainable ternary layer** — float weights stored and trained, quantized to `{-1, 0, +1}` only during the forward pass, with gradients flowing back through a Straight-Through Estimator (STE). Without STE the layer is untrainable (`jnp.sign` has zero gradient).

2. **Beta scaling** — `beta = mean(|W|)` multiplied into the output restores activation scale lost by quantization (BitNet §3.1). Without it, the layer has systematically smaller outputs than the float baseline.

## Proposed API (NNX)

```python
class TernaryDense(nnx.Module):
    def __init__(self, in_features, out_features, rngs, threshold=None):
        ...

    def __call__(self, x):
        abs_k = jnp.abs(self.kernel.value)
        t = 0.5 * jnp.mean(abs_k) if self.threshold is None else self.threshold
        k_q = jnp.sign(self.kernel.value) * (abs_k > t)
        # STE: forward uses k_q, backward sees identity
        k = self.kernel.value + jax.lax.stop_gradient(k_q - self.kernel.value)
        beta = jnp.mean(abs_k) if self.threshold is None else 1.0
        return beta * (x @ k) + (self.bias.value if self.bias else 0)
```

## What I'd like to know before writing the PR

- Is `TernaryDense` a fit for NNX, or is this better served by a standalone recipe/example in the docs?
- If it belongs in NNX, is there a preferred pattern for custom quantization layers (e.g. via `nnx.Variable` subclasses)?
- Any existing precedent in NNX for quantization-aware training layers would help me match the style.

Happy to write the implementation once the approach is agreed — or to contribute it as a documented example if that is preferred.

cc @vfdev-5 (based on feedback on #5466)

**Comment by samanklesaria** [comment]
Hi @eriirfos-eng - in an effort to limit the maintenance burden for new flax features, we don't want to add new layer types to the library unless their use is extremely widespread. While the results in the BitNet paper are cool, we'd want to see it become as omnipresent as MHA before we add it to nnx.

**Comment by simeon-kepp** [comment]
Hi @samanklesaria — appreciate the honest answer. Understood on the maintenance-burden concern. Let me push back with some concrete evidence before this closes.

**We are running this in production.** albert. (our research model) is a 26-layer dual-stream MoE, ~194.4M parameters, 2,044 tensors, trained exclusively with ternary weights from epoch 0. Not a proof-of-concept — ep4234 of continuous training, chip ATL 8.6852 nats. The only reason this model exists is TernaryDense + STE + beta-scaling, exactly as proposed.

**Inference throughput on CPU (from our latest bench, ep4234):**
| prompt type | speed |
|---|---|
| English prose | 52.6 tok/s |
| German (inflected) | 40–67 tok/s |
| code-adjacent | 62.5 tok/s |

This is on a laptop CPU, no GPU. The 55–65% zero-weight sparsity makes ternary inference fundamentally different from dense — you skip multiply-adds for the majority of weights. That advantage compounds with depth in a way no post-hoc quantization scheme matches.

**Maintenance:** I'm offering to own this layer completely — implementation, reviews, documentation, bug fixes. The cost to the Flax team is zero. I will maintain it for as long as Flax exists.

If NNX core is still off the table, I'd accept a first-class recipe in `flax/examples/` or the docs with proper STE and beta-scaling so people don't have to rediscover this from the paper. But I believe TernaryDense belongs in NNX proper — it's already the only viable path to efficient on-device inference at this depth.
